File: service/db/memory.py
import io
import json
from datetime import datetime

from service.db.contracts import DatabaseProvider
from service.logging import BackendLogger

db_logger = BackendLogger("db")

class InMemoryDB(DatabaseProvider):
    def __init__(self):
        self.users = {}
        self.history = []
        self.visualizations = {}
        self.genomes = []
        self.training_batches = {}
        self.research_visualizations = {}
        self.next_visualization_id = 1
        self.player_snapshots = []
        self.work_snapshots = []
        self.trade_snapshots = []
        self.night_snapshots = []

    def create_user(self, user_uuid: str, consent_agreed: bool) -> bool:
        self.users[user_uuid] = {
            "consent_agreed": consent_agreed,
            "created_at": datetime.now()
        }
        return True

    def user_exists(self, user_uuid: str) -> bool:
        return user_uuid in self.users

    def initialize_database(self):
        db_logger.info(
            "InMemoryDB ready. (Note: Data wipes on container restart)")
        
    def delete_research_visualizations(self, scope_type: str, scope_id: str):
        self.research_visualizations = {
            key: value
            for key, value in self.research_visualizations.items()
            if not (
                value["scope_type"] == scope_type
                and value["scope_id"] == scope_id
            )
        }

    def store_game_result(self, game_id, day_num, phase, snapshot_json,
                          training_batch_id=None, training_generation=None,
                          game_type=None, trade_count=None,
                          contest_count=None, lie_count=None):
        if isinstance(snapshot_json, str):
            data = json.loads(snapshot_json)
        else:
            data = snapshot_json

        self.history.append({
            "game_id": game_id,
            "day_num": day_num,
            "phase": phase,
            "data": data,
            "training_batch_id": training_batch_id,
            "training_generation": training_generation,
            "game_type": game_type or ("training" if training_batch_id else "human"),
            "trade_count": trade_count,
            "contest_count": contest_count,
            "lie_count": lie_count,
            "created_at": datetime.now()
        })

    def store_game_snapshot(self, game_id: str, day_num: int, phase: str, snapshot_json: str):
        self.history.append({
            "game_id": game_id,
            "day_num": day_num,
            "phase": phase,
            "data": json.loads(snapshot_json),
            "created_at": datetime.now()
        })

    def get_all_games(self):
        return sorted(self.history, key=lambda x: x['created_at'], reverse=True)

    def get_all_game_history(self) -> list:
        return sorted(self.history, key=lambda x: x['created_at'], reverse=True)

    def store_visualization(self, game_id: str, plot_name: str, figure):
        buf = io.BytesIO()
        figure.savefig(buf, format='png')
        buf.seek(0)

        if game_id not in self.visualizations:
            self.visualizations[game_id] = {}
        self.visualizations[game_id][plot_name] = buf.read()

    def store_player_snapshot(self, game_id, day_num, phase, player):
        self.player_snapshots.append({
            "game_id": game_id, "day_num": day_num, "phase": phase, "player": player
        })

    def store_work_snapshot(self, snapshot):
        self.work_snapshots.append(snapshot)

    def store_trade_snapshot(self, snapshot):
        self.trade_snapshots.append(snapshot)

    def store_night_snapshot(self, snapshot):
        self.night_snapshots.append(snapshot)

    def store_genome(self, name: str, shorthand: str, genome_json: str):
        self.genomes.append({
            "name": name,
            "shorthand_name": shorthand,
            "genome_data": json.loads(genome_json),
            "created_at": datetime.now()
        })

    def get_all_genomes(self) -> list:
        return sorted(self.genomes, key=lambda x: x['created_at'], reverse=True)

    def create_training_batch(self, batch_id: str, config: dict) -> bool:
        now = datetime.now()
        self.training_batches[batch_id] = {
            "batch_id": batch_id,
            "status": "running",
            "ruleset": config.get("ruleset"),
            "bot_model": config.get("bot_model"),
            "bot_count": config.get("bot_count"),
            "total_generations": config.get("total_generations"),
            "current_generation": 0,
            "current_game_id": None,
            "started_at": now,
            "completed_at": None,
            "last_heartbeat_at": now,
            "phase": "pending",
            "last_error": None,
            "base_genome_id": config.get("base_genome_id"),
            "final_champion_genome_id": None,
            "config": config.get("config", {}),
            "games_per_generation": (config.get("config", {}) or {}).get(
                "games_per_generation"),
            "games_completed": 0,
            "games_failed": 0,
            "generation_statistics": [],
            "games": [],
        }
        return True

    def mark_training_batch_game_started(self, batch_id: str, game_id: str,
                                         generation: int,
                                         attempt: int | None = None):
        batch = self.training_batches.get(batch_id)
        if not batch:
            return
        batch["status"] = "running"
        batch["current_game_id"] = game_id
        batch["current_generation"] = generation
        if not isinstance(batch.get("games"), list):
            batch["games"] = []
        batch["games"].append({
            "game_id": game_id,
            "generation": generation,
            "attempt": attempt,
            "status": "spawning",
            "error_message": None,
            "genome_count": 0,
            "best_fitness": None,
            "average_fitness": None,
        })

    def mark_training_batch_game_running(self, batch_id: str, game_id: str):
        batch = self.training_batches.get(batch_id)
        if not batch:
            return
        for game in batch.get("games", []):
            if game.get("game_id") == game_id:
                game["status"] = "running"
                return

    def mark_training_batch_game_failed(self, batch_id: str, game_id: str,
                                        error_message: str):
        batch = self.training_batches.get(batch_id)
        if not batch:
            return
        for game in batch.get("games", []):
            if game.get("game_id") == game_id:
                if game.get("status") != "failed":
                    batch["games_failed"] = int(batch.get("games_failed", 0) or 0) + 1
                game["status"] = "failed"
                game["error_message"] = error_message
                game["genome_count"] = int(game.get("genome_count", 0) or 0)
                batch["games_completed"] = len([
                    entry for entry in batch.get("games", [])
                    if entry.get("status") in ("completed", "failed", "skipped")
                ])
                return

    def mark_training_batch_game_completed(self, batch_id: str, game_id: str,
                                           genome_count: int,
                                           fitness_summary: dict | None = None):
        batch = self.training_batches.get(batch_id)
        if not batch:
            return
        summary = fitness_summary or {}
        for game in batch.get("games", []):
            if game.get("game_id") == game_id:
                game["status"] = "completed"
                game["error_message"] = None
                game["genome_count"] = int(genome_count or 0)
                game["best_fitness"] = summary.get("best_fitness")
                game["average_fitness"] = summary.get("average_fitness")
                batch["games_completed"] = len([
                    entry for entry in batch.get("games", [])
                    if entry.get("status") in ("completed", "failed", "skipped")
                ])
                batch["games_failed"] = len([
                    entry for entry in batch.get("games", [])
                    if entry.get("status") == "failed"
                ])
                return

    def record_training_batch_heartbeat(self, batch_id: str, phase: str,
                                        current_generation: int,
                                        current_game_id: str | None = None):
        batch = self.training_batches.get(batch_id)
        if not batch:
            return
        batch["last_heartbeat_at"] = datetime.now()
        batch["phase"] = phase
        batch["current_generation"] = current_generation
        batch["current_game_id"] = current_game_id

    def update_training_batch_status(self, batch_id: str, status: str,
                                     error_message: str | None = None):
        batch = self.training_batches.get(batch_id)
        if not batch:
            return
        batch["status"] = status
        batch["last_error"] = error_message
        if status in ("completed", "failed", "cancelled", "stalled"):
            batch["completed_at"] = datetime.now()

    def append_training_batch_generation_stats(self, batch_id: str, stats: dict):
        batch = self.training_batches.get(batch_id)
        if batch:
            batch.setdefault("generation_statistics", []).append(stats)

    def complete_training_batch(self, batch_id: str,
                                final_champion_genome_id: str | None = None):
        batch = self.training_batches.get(batch_id)
        if not batch:
            return
        batch["status"] = "completed"
        batch["completed_at"] = datetime.now()
        batch["final_champion_genome_id"] = final_champion_genome_id

    def get_training_batches(self) -> list:
        return sorted(
            [dict(batch) for batch in self.training_batches.values()],
            key=lambda batch: batch["started_at"], reverse=True)

    def get_training_batch(self, batch_id: str) -> dict | None:
        batch = self.training_batches.get(batch_id)
        return dict(batch) if batch else None

    def get_training_games(self, batch_id: str) -> list:
        batch = self.training_batches.get(batch_id)
        if not batch:
            return []
        return [dict(game) for game in batch.get("games", [])]

    def store_research_visualization(self, scope_type: str, scope_id: str,
                                     name: str, title: str, mime_type: str,
                                     image_bytes: bytes,
                                     metadata: dict | None = None):
        visualization_id = str(self.next_visualization_id)
        self.next_visualization_id += 1
        self.research_visualizations[visualization_id] = {
            "id": visualization_id,
            "scope_type": scope_type,
            "scope_id": scope_id,
            "name": name,
            "title": title,
            "mime_type": mime_type,
            "image_bytes": image_bytes,
            "metadata": metadata or {},
            "created_at": datetime.now(),
        }
        return visualization_id

    def get_research_visualizations(self, scope_type: str, scope_id: str) -> list:
        visualizations = []
        for visualization in self.research_visualizations.values():
            if (visualization["scope_type"] == scope_type and
                    visualization["scope_id"] == scope_id):
                item = dict(visualization)
                item.pop("image_bytes", None)
                item["url"] = f"/api/research/visualizations/{item['id']}"
                visualizations.append(item)
        return sorted(visualizations, key=lambda item: item["created_at"])

    def get_research_visualization(self, visualization_id):
        visualization = self.research_visualizations.get(str(visualization_id))
        return dict(visualization) if visualization else None


# --- 3. The MySQL Provider (Prod) ---

File: service/src/db/memory.ts
import type { JsonValue } from "@takes-a-village/shared";

import type {
  DatabaseProvider,
  GameRecord,
  GameResultOptions,
  GenomeRecord,
  JsonObject,
  StoreVisualizationInput,
  TrainingBatchConfig,
  TrainingBatchRecord,
  TrainingGameRecord,
  VisualizationRecord,
} from "./contracts.js";

function clone<T>(value: T): T {
  return structuredClone(value);
}

export class MemoryDatabase implements DatabaseProvider {
  private readonly users = new Map<string, { consentAgreed: boolean; createdAt: Date }>();
  private readonly history: GameRecord[] = [];
  private readonly games: GameRecord[] = [];
  private readonly genomes: GenomeRecord[] = [];
  private readonly trainingBatches = new Map<string, TrainingBatchRecord>();
  private readonly visualizations = new Map<string, VisualizationRecord>();
  readonly playerSnapshots: JsonValue[] = [];
  readonly workSnapshots: JsonValue[] = [];
  readonly tradeSnapshots: JsonValue[] = [];
  readonly nightSnapshots: JsonValue[] = [];
  private nextVisualizationId = 1;

  async initialize(): Promise<void> {}
  async close(): Promise<void> {}

  async createUser(userId: string, consentAgreed = true): Promise<boolean> {
    this.users.set(userId, { consentAgreed, createdAt: new Date() });
    return true;
  }

  async userExists(userId: string): Promise<boolean> {
    return this.users.has(userId);
  }

  async storeGameSnapshot(gameId: string, dayNum: number, phase: string, snapshot: JsonValue): Promise<void> {
    this.history.push({ game_id: gameId, day_num: dayNum, phase, data: clone(snapshot), created_at: new Date() });
  }

  async storeGameResult(gameId: string, dayNum: number, phase: string, snapshot: JsonValue, options: GameResultOptions = {}): Promise<void> {
    const row: GameRecord = {
      game_id: gameId,
      day_num: dayNum,
      phase,
      data: clone(snapshot),
      created_at: new Date(),
      game_type: options.gameType ?? (options.trainingBatchId ? "training" : "human"),
      training_batch_id: options.trainingBatchId ?? null,
      training_generation: options.trainingGeneration ?? null,
      trade_count: options.tradeCount ?? null,
      contest_count: options.contestCount ?? null,
      lie_count: options.lieCount ?? null,
    };
    this.games.push(row);
    this.history.push(row);
  }

  async getAllGameHistory(): Promise<GameRecord[]> {
    return clone([...this.history].reverse());
  }

  async getAllGames(): Promise<GameRecord[]> {
    return clone([...this.games].reverse());
  }

  async storePlayerSnapshot(gameId: string, dayNum: number, phase: string, player: JsonValue): Promise<void> {
    this.playerSnapshots.push({ game_id: gameId, day_num: dayNum, phase, player } as JsonValue);
  }
  async storeWorkSnapshot(snapshot: JsonValue): Promise<void> { this.workSnapshots.push(clone(snapshot)); }
  async storeTradeSnapshot(snapshot: JsonValue): Promise<void> { this.tradeSnapshots.push(clone(snapshot)); }
  async storeNightSnapshot(snapshot: JsonValue): Promise<void> { this.nightSnapshots.push(clone(snapshot)); }

  async storeGenome(name: string, shorthand: string, genome: JsonValue): Promise<void> {
    this.genomes.push({ id: this.genomes.length + 1, name, shorthand_name: shorthand, genome_data: clone(genome), created_at: new Date() });
  }

  async getAllGenomes(): Promise<GenomeRecord[]> { return clone([...this.genomes].reverse()); }

  async createTrainingBatch(batchId: string, config: TrainingBatchConfig): Promise<boolean> {
    const now = new Date();
    const nestedConfig = clone(config.config ?? {});
    this.trainingBatches.set(batchId, {
      ...clone(config),
      batch_id: batchId,
      status: "running",
      current_generation: 0,
      current_game_id: null,
      started_at: now,
      completed_at: null,
      last_heartbeat_at: now,
      phase: "pending",
      last_error: null,
      final_champion_genome_id: null,
      config: nestedConfig,
      generation_statistics: [],
      games: [],
      games_per_generation: typeof nestedConfig.games_per_generation === "number" ? nestedConfig.games_per_generation : undefined,
      games_completed: 0,
      games_failed: 0,
    });
    return true;
  }

  private findBatch(batchId: string): TrainingBatchRecord | undefined { return this.trainingBatches.get(batchId); }

  async markTrainingBatchGameStarted(batchId: string, gameId: string, generation: number, attempt: number | null = null): Promise<void> {
    const batch = this.findBatch(batchId);
    if (!batch) return;
    batch.status = "running";
    batch.current_game_id = gameId;
    batch.current_generation = generation;
    batch.games.push({ game_id: gameId, generation, attempt, status: "spawning", error_message: null, genome_count: 0, best_fitness: null, average_fitness: null });
  }

  async markTrainingBatchGameRunning(batchId: string, gameId: string): Promise<void> {
    const game = this.findBatch(batchId)?.games.find((candidate) => candidate.game_id === gameId);
    if (game) game.status = "running";
  }

  async markTrainingBatchGameFailed(batchId: string, gameId: string, errorMessage: string): Promise<void> {
    const batch = this.findBatch(batchId);
    const game = batch?.games.find((candidate) => candidate.game_id === gameId);
    if (!batch || !game) return;
    game.status = "failed";
    game.error_message = errorMessage;
    this.updateGameCounts(batch);
  }

  async markTrainingBatchGameCompleted(batchId: string, gameId: string, genomeCount: number, fitnessSummary: JsonObject = {}): Promise<void> {
    const batch = this.findBatch(batchId);
    const game = batch?.games.find((candidate) => candidate.game_id === gameId);
    if (!batch || !game) return;
    game.status = "completed";
    game.error_message = null;
    game.genome_count = genomeCount;
    game.best_fitness = typeof fitnessSummary.best_fitness === "number" ? fitnessSummary.best_fitness : null;
    game.average_fitness = typeof fitnessSummary.average_fitness === "number" ? fitnessSummary.average_fitness : null;
    this.updateGameCounts(batch);
  }

  private updateGameCounts(batch: TrainingBatchRecord): void {
    batch.games_completed = batch.games.filter((game) => ["completed", "failed", "skipped"].includes(game.status)).length;
    batch.games_failed = batch.games.filter((game) => game.status === "failed").length;
  }

  async recordTrainingBatchHeartbeat(batchId: string, phase: string, currentGeneration: number, currentGameId: string | null = null): Promise<void> {
    const batch = this.findBatch(batchId);
    if (!batch) return;
    batch.last_heartbeat_at = new Date();
    batch.phase = phase;
    batch.current_generation = currentGeneration;
    batch.current_game_id = currentGameId;
  }

  async updateTrainingBatchStatus(batchId: string, status: string, errorMessage: string | null = null): Promise<void> {
    const batch = this.findBatch(batchId);
    if (!batch) return;
    batch.status = status;
    batch.last_error = errorMessage;
    if (["completed", "failed", "cancelled", "stalled"].includes(status)) batch.completed_at = new Date();
  }

  async appendTrainingBatchGenerationStats(batchId: string, stats: JsonObject): Promise<void> {
    this.findBatch(batchId)?.generation_statistics.push(clone(stats));
  }

  async completeTrainingBatch(batchId: string, finalChampionGenomeId: string | null = null): Promise<void> {
    const batch = this.findBatch(batchId);
    if (!batch) return;
    batch.status = "completed";
    batch.completed_at = new Date();
    batch.final_champion_genome_id = finalChampionGenomeId;
  }

  async getTrainingBatches(): Promise<TrainingBatchRecord[]> {
    return clone([...this.trainingBatches.values()].sort((a, b) => b.started_at.getTime() - a.started_at.getTime()));
  }
  async getTrainingBatch(batchId: string): Promise<TrainingBatchRecord | null> { return clone(this.findBatch(batchId) ?? null); }
  async getTrainingGames(batchId: string): Promise<TrainingGameRecord[]> { return clone(this.findBatch(batchId)?.games ?? []); }

  async storeResearchVisualization(input: StoreVisualizationInput): Promise<string> {
    const id = String(this.nextVisualizationId++);
    this.visualizations.set(id, {
      id,
      scope_type: input.scopeType,
      scope_id: input.scopeId,
      name: input.name,
      title: input.title,
      mime_type: input.mimeType,
      image_bytes: Buffer.from(input.imageBytes),
      metadata: clone(input.metadata ?? {}),
      created_at: new Date(),
    });
    return id;
  }

  async getResearchVisualizations(scopeType: string, scopeId: string): Promise<VisualizationRecord[]> {
    return [...this.visualizations.values()]
      .filter((item) => item.scope_type === scopeType && item.scope_id === scopeId)
      .sort((a, b) => a.created_at.getTime() - b.created_at.getTime())
      .map(({ image_bytes: _imageBytes, ...item }) => clone({ ...item, url: `/api/research/visualizations/${item.id}` }));
  }

  async getResearchVisualization(visualizationId: string): Promise<VisualizationRecord | null> {
    const item = this.visualizations.get(visualizationId);
    return item ? { ...clone(item), image_bytes: item.image_bytes ? Buffer.from(item.image_bytes) : undefined } : null;
  }

  async deleteResearchVisualizations(scopeType: string, scopeId: string): Promise<void> {
    for (const [id, item] of this.visualizations) {
      if (item.scope_type === scopeType && item.scope_id === scopeId) this.visualizations.delete(id);
    }
  }
}
